dds/src/Conversion.cpp

```diff
@@ -30,144 +29,744 @@ using namespace eprosima::fastrtps;
+            break;
+            case ::xtypes::TypeKind::STRING_TYPE:
+            {
+                std::string value;
+                ret = from->get_string_value(value, id);
+                to.push(value);
+            }
+            break;
+            case ::xtypes::TypeKind::WSTRING_TYPE:
+            {
+                std::wstring value;
+                ret = from->get_wstring_value(value, id);
+                to.push(value);
+            }
+            break;
+            case ::xtypes::TypeKind::ARRAY_TYPE:
+            {
+                DynamicData* array = from->loan_value(id);
+                ::xtypes::DynamicData soss_array(type.content_type());
+                set_array_data(array, soss_array.ref(), std::vector<uint32_t>());
+                from->return_loaned_value(array);
+                to.push(soss_array);
+                ret = ResponseCode::RETCODE_OK;
+                break;
+            }
+            case ::xtypes::TypeKind::SEQUENCE_TYPE:
+            {
+                DynamicData* seq = from->loan_value(id);
+                ::xtypes::DynamicData soss_seq(type.content_type());
+                set_sequence_data(seq, soss_seq.ref());
+                from->return_loaned_value(seq);
+                to.push(soss_seq);
+                ret = ResponseCode::RETCODE_OK;
+                break;
+            }
+            case ::xtypes::TypeKind::STRUCTURE_TYPE:
+            {
+                DynamicData* st = from->loan_value(id);
+                ::xtypes::DynamicData soss_st(type.content_type());
+                set_struct_data(st, soss_st.ref());
+                from->return_loaned_value(st);
+                to.push(soss_st);
+                ret = ResponseCode::RETCODE_OK;
+                break;
+            }
+            default:
+            {
+                std::stringstream ss;
+                ss << "Unexpected data type: " << to.type().name() << ".";
+                throw DDSMiddlewareException(ss.str());
+            }
+        }
+
+        if (ret != ResponseCode::RETCODE_OK)
+        {
+            std::stringstream ss;
+            ss << "Error parsing from dynamic type '" << to.type().name();
+            throw DDSMiddlewareException(ss.str());
+        }
+    }
+}
+
+void Conversion::set_array_data(
+        const DynamicData* c_from,
+        ::xtypes::WritableDynamicDataRef to,
+        const std::vector<uint32_t>& indexes)
+{
+    const ::xtypes::ArrayType& type = static_cast<const ::xtypes::ArrayType&>(to.type());
+    const ::xtypes::DynamicType& inner_type = type.content_type();
+    DynamicData* from = const_cast<DynamicData*>(c_from);
+    MemberId id;
+
+    for (uint32_t idx = 0; idx < type.dimension(); ++idx)
+    {
+        std::vector<uint32_t> new_indexes = indexes;
+        new_indexes.push_back(idx);
+        ResponseCode ret = ResponseCode::RETCODE_ERROR;
+        switch (inner_type.kind())
+        {
+            case ::xtypes::TypeKind::BOOLEAN_TYPE:
+            {
```

Comment:
Could be a macro? (see similar occurences)

Reply:
Same as above.